### DISCUSSION TITLE
How to access custom attributes from implements section via reflection?

### DISCUSSION THREAD

**Discussion by n9** [dotnet/runtime#49900] [Q&A]
I want to access information about nullable reference types (https://github.com/dotnet/runtime/issues/29723). Some information seems to be hidden in custom attributes of implements section (not sure how it is called officially). How to access the following attribute via reflection?

```
.class private auto ansi beforefieldinit BarN
    extends [System.Private.CoreLib]System.Object
    implements .custom instance void System.Runtime.CompilerServices.NullableAttribute::.ctor(uint8[]) = (
        01 00 02 00 00 00 00 02 00 00
    )
    class IBar`1<string>
```

Context is here: https://github.com/RicoSuter/Namotion.Reflection/issues/67.

**Comment by Joe4evr** [comment]
The entire point of #29723 is so that *you* don't have to do this in the first place. You should wait until that API is implemented and released, *then* you can query for the nullability information with normal reflection code.

**Reply by n9** [reply]
@Joe4evr Yes, if I could wait, I didn't have to do this:) 

Do you know, how to access the custom attributes from implements section?

**Reply by Joe4evr** [reply]
> Do you know, how to access the custom attributes from implements section?

That's as simple as `type.GetCustomAttributesData()`.

**Reply by n9** [reply]
This won't return the custom attributes **from implements section**. Try it.

**Reply by huoyaoyuan** [reply]
You can see the PR #54985

**Reply by Joe4evr** [reply]
> This won't return the custom attributes **from implements section**. Try it.

Then why does the PR use it successfully?
https://github.com/dotnet/runtime/blob/bfd9f1b8b139a5138c185e1e961bfe79051013a6/src/libraries/System.Private.CoreLib/src/System/Reflection/NullabilityInfoContext.cs#L37

**Reply by n9** [reply]
@Joe4evr The line, you posted, access custom attributes from members, but not the attributes from the implements section from the question:

```
.class private auto ansi beforefieldinit BarN
    extends [System.Private.CoreLib]System.Object
    implements .custom instance void System.Runtime.CompilerServices.NullableAttribute::.ctor(uint8[]) = (
        01 00 02 00 00 00 00 02 00 00
    )
    class IBar`1<string>
```

The code in the linked issue is following:

```C#
interface IBar<T> { }

class Bar : IBar<string> { }
class BarN : IBar<string?> { }
```

**Reply by n9** [reply]
@huoyaoyuan What part of the PR resolves nullability of generic argument in `BarN`.
